File: Bukkit/src/main/java/me/mrkirby153/kcutils/scoreboard/ScoreboardTeam.java
package me.mrkirby153.kcutils.scoreboard;

import net.md_5.bungee.api.ChatColor;
import org.bukkit.entity.Player;

import java.util.HashSet;
import java.util.Set;
import java.util.UUID;


public class ScoreboardTeam {

    protected String teamName;

    protected Set<UUID> players = new HashSet<>();

    protected ChatColor color;
    protected ChatColor prefixColor;

    protected String prefixFormat = "[%s]";

    protected boolean showPrefix = false;
    protected String prefix = null;

    protected boolean friendlyFire = false;
    protected boolean seeInvisible = false;

    public ScoreboardTeam(String teamName, ChatColor color) {
        this.teamName = teamName;
        this.color = color;
        this.prefixColor = color;
    }

    public void addPlayer(Player player) {
        players.add(player.getUniqueId());
    }

    public boolean canSeeInvisibles() {
        return seeInvisible;
    }

    @Override
    public boolean equals(Object obj) {
        return obj instanceof ScoreboardTeam && ((ScoreboardTeam) obj).getTeamName().equals(this.getTeamName());
    }

    public ChatColor getColor() {
        return color;
    }

    public String getFilteredName() {
        return getTeamName().substring(0, Math.min(getTeamName().length(), 16));
    }

    public Set<UUID> getPlayers() {
        return players;
    }

    public String getPrefix() {
        return prefix;
    }

    public void setPrefix(String prefix) {
        this.prefix = prefix;
    }

    public ChatColor getPrefixColor() {
        return prefixColor;
    }

    public void setPrefixColor(ChatColor prefixColor) {
        this.prefixColor = prefixColor;
    }

    public String getPrefixFormat() {
        return prefixFormat;
    }

    public void setPrefixFormat(String prefixFormat) {
        this.prefixFormat = prefixFormat;
    }

    public String getTeamName() {
        return teamName;
    }

    public void setTeamName(String teamName) {
        this.teamName = teamName;
    }

    public boolean isFriendlyFire() {
        return friendlyFire;
    }

    public void setFriendlyFire(boolean friendlyFire) {
        this.friendlyFire = friendlyFire;
    }

    public void removePlayer(Player player) {
        players.remove(player.getUniqueId());
    }

    public void removePlayer(UUID uuid){
        players.remove(uuid);
    }

    public void setSeeInvisible(boolean seeInvisible) {
        this.seeInvisible = seeInvisible;
    }

    public void setShowPrefix(boolean showPrefix) {
        this.showPrefix = showPrefix;
    }

    public boolean showPrefix() {
        return showPrefix;
    }
}

File: Bukkit/src/main/kotlin/me/mrkirby153/kcutils/scoreboard/ScoreboardTeam.kt
package me.mrkirby153.kcutils.scoreboard

import net.kyori.adventure.text.format.NamedTextColor
import org.bukkit.entity.Player
import java.util.UUID

/**
 * A team on a [Scoreboard][KirbyScoreboard]
 *
 * @param teamName  The name of the team
 * @param color     The color of the team
 */
open class ScoreboardTeam(val teamName: String, var color: NamedTextColor) {

    /**
     * A list of players on the scoreboard
     */
    val players = mutableSetOf<UUID>()

    var prefixColor = color

    /**
     * The team name truncated to 16 chars, used for internal reference
     */
    val filteredName: String
        get() = teamName.substring(0, Math.min(teamName.length, 16))

    /**
     * If the team's prefix should be shown
     */
    var showPrefix = false

    /**
     * The prefix to show
     */
    var prefix: String? = null

    /**
     * The format of the prefix
     */
    var prefixFormat = "[%s]"

    /**
     * If friendly fire is enabled
     */
    var friendlyFire = false

    /**
     * If the players of the team can see each other if invisible
     */
    var seeInvisible = false

    override fun equals(other: Any?): Boolean {
        return other is ScoreboardTeam && other.teamName == this.teamName
    }

    override fun hashCode(): Int {
        var result = teamName.hashCode()
        result = 31 * result + color.hashCode()
        result = 31 * result + prefixColor.hashCode()
        result = 31 * result + showPrefix.hashCode()
        result = 31 * result + (prefix?.hashCode() ?: 0)
        result = 31 * result + friendlyFire.hashCode()
        result = 31 * result + seeInvisible.hashCode()
        return result
    }

    /**
     * Removes a player from the team
     *
     * @param player The player to remove
     */
    open fun removePlayer(player: Player) {
        players.remove(player.uniqueId)
    }

    /**
     * Adds a player to the team
     *
     * @param player The player to add to the team
     */
    open fun addPlayer(player: Player) {
        players.add(player.uniqueId)
    }

    /**
     * Removes a player from the team
     *
     * @param uuid  The user to remove from the team
     */
    fun removePlayer(uuid: UUID) {
        players.remove(uuid)
    }


    /**
     * Adds a player to the team
     *
     * @param uuid The user to add to the team
     */
    fun addPlayer(uuid: UUID) {
        players.add(uuid)
    }
}